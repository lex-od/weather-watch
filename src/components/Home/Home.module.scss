@use "styles/shared.scss";

.loadingBackdrop {
  position: absolute;
  top: 24px;
  left: 24px;
  right: 0;
  bottom: 0;
  z-index: 10;

  display: flex;
  justify-content: center;
  align-items: center;

  backdrop-filter: blur(4px);
}
