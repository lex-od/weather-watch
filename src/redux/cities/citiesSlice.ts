import { createSlice } from "@reduxjs/toolkit";
import type { PayloadAction } from "@reduxjs/toolkit";
import { RootState } from "redux/store";

export interface ICitiesState {
  list: null;
}

const initialState: ICitiesState = {
  list: null,
};

export const citiesSlice = createSlice({
  name: "cities",
  initialState,
  reducers: {
    addNewCity: (state, action: PayloadAction<any>) => {
      //
    },
  },
});

export const { addNewCity } = citiesSlice.actions;
export default citiesSlice.reducer;

// ðŸ“Œ Selectors

export const getCitiesList = (state: RootState) => state.cities.list;
