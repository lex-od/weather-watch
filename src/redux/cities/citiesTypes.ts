import { IGetCurrentWeatherResponse } from "services/api/types";

export interface ICitiesState {
  selectedCities: ICityItem[];
}

export interface ICityItem {
  id: number;
  name: string;
  state: string;
  country: string;
  coord: {
    lon: number;
    lat: number;
  };
}

// City with weather

export interface ICityWithWeatherItem extends ICityItem {
  weather: IGetCurrentWeatherResponse | null;
}

// Add new city

export interface IAddCityPayload {
  newCity: ICityItem;
}

// Delete city

export interface IDeleteCityPayload {
  cityId: number;
}
