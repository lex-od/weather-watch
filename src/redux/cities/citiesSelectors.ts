import { createSelector } from "@reduxjs/toolkit";
import { RootState } from "redux/store";
import weatherSelectors from "redux/weather/weatherSelectors";
import { IGetCurrentWeatherResponse } from "services/api/types";
import { ICityItem, ICityWithWeatherItem } from "./citiesTypes";

const getSelectedCities = (state: RootState) => state.cities.selectedCities;

const getCityByIdWithWeather = createSelector(
  [
    getSelectedCities,
    weatherSelectors.getWeatherByCities,
    (_, cityId: number) => cityId,
  ],
  (
    selectedCities: ICityItem[],
    weatherByCities: IGetCurrentWeatherResponse[],
    cityId
  ): ICityWithWeatherItem | null => {
    const city = selectedCities.find(({ id }) => id === cityId);
    if (!city) return null;

    const weather = weatherByCities.find(({ id }) => id === cityId);
    return {
      ...city,
      weather: weather || null,
    };
  }
);

const getSelectedCitiesWithWeather = createSelector(
  [getSelectedCities, weatherSelectors.getWeatherByCities],
  (
    selectedCities: ICityItem[],
    weatherByCities: IGetCurrentWeatherResponse[]
  ): ICityWithWeatherItem[] => {
    return selectedCities.map((city) => {
      const weather = weatherByCities.find(({ id }) => id === city.id);
      return {
        ...city,
        weather: weather || null,
      };
    });
  }
);

const getSelectedCityIdsWithNoWeather = createSelector(
  [getSelectedCities, weatherSelectors.getWeatherByCities],
  (
    selectedCities: ICityItem[],
    weatherByCities: IGetCurrentWeatherResponse[]
  ): number[] => {
    const noWeatherCities = selectedCities.filter((city) => {
      return weatherByCities.every(({ id }) => id !== city.id);
    });
    return noWeatherCities.map(({ id }) => id);
  }
);

const citiesSelectors = {
  getSelectedCities,
  getCityByIdWithWeather,
  getSelectedCitiesWithWeather,
  getSelectedCityIdsWithNoWeather,
};
export default citiesSelectors;
