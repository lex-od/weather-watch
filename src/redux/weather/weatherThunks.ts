import { createAsyncThunk } from "@reduxjs/toolkit";

import { api } from "services";
import { IGetCurrentWeatherParams } from "services/api/types";
import { showError } from "utils";
import citiesSelectors from "redux/cities/citiesSelectors";
import { RootState } from "redux/store";

const getCurrentWeather = createAsyncThunk(
  "weather/getCurrentWeather",
  async (payload: IGetCurrentWeatherParams) => {
    try {
      const { data } = await api.getCurrentWeather(payload);
      return data;
    } catch (error) {
      showError(error);
      throw error;
    }
  }
);

const getCurrentWeatherBySelectedCities = createAsyncThunk(
  "weather/getCurrentWeatherBySelectedCities",
  async (_, { getState }) => {
    const noWeatherIds = citiesSelectors.getSelectedCityIdsWithNoWeather(
      getState() as RootState
    );

    try {
      const results = await Promise.all(
        noWeatherIds.map((id) => api.getCurrentWeather({ id }))
      );
      return results.map(({ data }) => data);
    } catch (error) {
      showError(error);
      throw error;
    }
  }
);

const weatherThunks = {
  getCurrentWeather,
  getCurrentWeatherBySelectedCities,
};
export default weatherThunks;
