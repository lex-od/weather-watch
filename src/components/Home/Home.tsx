import { useEffect, useState } from "react";
import Box from "@mui/material/Box";
import Grid from "@mui/material/Grid";
import CircularProgress from "@mui/material/CircularProgress";
import Button from "@mui/material/Button";
import { Typography } from "@mui/material";

import css from "./Home.module.scss";
import { useAppDispatch, useAppSelector } from "hooks";
import weatherThunks from "redux/weather/weatherThunks";
import citiesSls from "redux/cities/citiesSelectors";
import weatherSls from "redux/weather/weatherSelectors";
import { ICityItem } from "redux/cities/citiesTypes";
import { addNewCity } from "redux/cities/citiesSlice";
import { WeatherCard } from "./WeatherCard/WeatherCard";
import { AddCityDialog } from "./AddCityDialog/AddCityDialog";

export const Home = () => {
  const dispatch = useAppDispatch();
  const cities = useAppSelector(citiesSls.getSelectedCities);
  const citiesWithWeather = useAppSelector(
    citiesSls.getSelectedCitiesWithWeather
  );
  const remainingCities = useAppSelector(citiesSls.getRemainingAvailableCities);
  const loading = useAppSelector(weatherSls.getWeatherAllLoading);

  const [isAddModal, setIsAddModal] = useState(false);

  const closeAddModal = () => setIsAddModal(false);

  const handleAddCitySubmit = (newCity: ICityItem) => {
    dispatch(addNewCity({ newCity }));
    closeAddModal();
  };

  // Получаем только отсутствующую погоду
  useEffect(() => {
    dispatch(weatherThunks.getCurrentWeatherBySelectedCities());
  }, [dispatch, cities]);

  return (
    <Box sx={{ padding: "15px 0" }}>
      <Typography variant="h2" sx={{ textAlign: "center", mb: "50px" }}>
        Weather watch
      </Typography>

      <Grid container spacing={3} sx={{ mb: "50px", position: "relative" }}>
        {citiesWithWeather.map((item) => (
          <Grid key={item.id} item xs={4}>
            <WeatherCard item={item} />
          </Grid>
        ))}

        {loading && (
          <div className={css.loadingBackdrop}>
            <CircularProgress size={48} />
          </div>
        )}
      </Grid>

      <Box sx={{ textAlign: "center" }}>
        <Button
          variant="outlined"
          size="large"
          onClick={() => setIsAddModal(true)}
        >
          Add city
        </Button>
      </Box>

      <AddCityDialog
        cities={remainingCities}
        onSubmit={handleAddCitySubmit}
        open={isAddModal}
        onClose={closeAddModal}
      />
    </Box>
  );
};

// import { Link as RouterLink } from "react-router-dom";
// import Link from "@mui/material/Link";

// <Link
//   component={RouterLink}
//   to={routes.addCity}
//   variant="button"
//   underline="none"
//   className={css.addCityLink}
// >
//   Add city
// </Link>;
