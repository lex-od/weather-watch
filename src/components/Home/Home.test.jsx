import { render, screen, fireEvent } from "@testing-library/react";
import * as redux from "react-redux";
import * as router from "react-router-dom";
import weatherThunks from "redux/weather/weatherThunks";
import * as citiesActs from "redux/cities/citiesSlice";

import { Home } from "./Home";

const state = {
  cities: {
    selectedCities: [
      {
        id: 698740,
        name: "Odessa",
        state: "",
        country: "UA",
        coord: {
          lon: 30.732622,
          lat: 46.477474,
        },
      },
    ],
  },
  weather: {
    weatherByCities: [
      {
        coord: {
          lon: 30.7326,
          lat: 46.4775,
        },
        weather: [
          {
            id: 801,
            main: "Clouds",
            description: "few clouds",
            icon: "02d",
          },
        ],
        base: "stations",
        main: {
          temp: 1.88,
          feels_like: -1.45,
          temp_min: 1.88,
          temp_max: 1.88,
          pressure: 1021,
          humidity: 63,
          sea_level: 1021,
          grnd_level: 1014,
        },
        visibility: 10000,
        wind: {
          speed: 3.26,
          deg: 113,
          gust: 3.54,
        },
        clouds: {
          all: 11,
        },
        dt: 1671013015,
        sys: {
          country: "UA",
          sunrise: 1670995990,
          sunset: 1671027013,
        },
        timezone: 7200,
        id: 698740,
        name: "Odesa",
        cod: 200,
      },
    ],
    weatherAllLoading: false,
  },
};

const stateWeatherLoading = {
  ...state,
  weather: {
    weatherByCities: [],
    weatherAllLoading: true,
  },
};

const cherkasyCity = {
  id: 710791,
  name: "Cherkasy",
  state: "",
  country: "UA",
  coord: {
    lon: 32.062069,
    lat: 49.428539,
  },
};

jest.mock("react-redux");
jest.mock("react-router-dom");
const spyUseDispatch = jest.spyOn(redux, "useDispatch");
const spyUseSelector = jest.spyOn(redux, "useSelector");
const spyNavigate = jest.spyOn(router, "useNavigate");

describe("Home component", () => {
  it("Home renders", () => {
    spyUseDispatch.mockReturnValue(jest.fn());
    spyUseSelector.mockImplementation((cb) => cb(state));
    spyNavigate.mockReturnValue(jest.fn());

    render(<Home />);

    expect(screen.getByText("Weather watch")).toBeInTheDocument();
    expect(screen.getByText("Add city")).toBeInTheDocument();
  });

  it("Home with card renders", () => {
    spyUseDispatch.mockReturnValue(jest.fn());
    spyUseSelector.mockImplementation((cb) => cb(state));
    spyNavigate.mockReturnValue(jest.fn());

    render(<Home />);

    expect(screen.getByText("Weather watch")).toBeInTheDocument();
    expect(screen.getByText("Add city")).toBeInTheDocument();
    expect(screen.getByRole("img")).toBeInTheDocument();
    expect(screen.getByText(/Wind:/)).toBeInTheDocument();
    expect(screen.getByText("Update")).toBeInTheDocument();
    expect(screen.getByText("Delete")).toBeInTheDocument();
  });

  it("Loading renders", () => {
    spyUseDispatch.mockReturnValue(jest.fn());
    spyUseSelector.mockImplementation((cb) => cb(stateWeatherLoading));
    spyNavigate.mockReturnValue(jest.fn());

    render(<Home />);

    expect(screen.getByRole("progressbar")).toBeInTheDocument();
  });

  it("Should dispatch actions", () => {
    // Mock return values
    const mockDispatch = jest.fn();
    spyUseDispatch.mockReturnValue(mockDispatch);
    spyUseSelector.mockImplementation((cb) => cb(state));
    spyNavigate.mockReturnValue(jest.fn());

    // Spy actions
    const spyGetCurrentWeatherBySelectedCities = jest.spyOn(
      weatherThunks,
      "getCurrentWeatherBySelectedCities"
    );
    const spyAddNewCity = jest.spyOn(citiesActs, "addNewCity");

    render(<Home />);

    // Get weather by selected cities
    expect(mockDispatch).toHaveBeenCalledTimes(1);
    expect(spyGetCurrentWeatherBySelectedCities).toHaveBeenCalledTimes(1);

    // Add new city
    fireEvent.click(screen.getByText("Add city"));
    fireEvent.click(screen.getByText("Cherkasy"));
    fireEvent.click(screen.getByText("Add"));
    expect(mockDispatch).toHaveBeenCalledTimes(2);
    expect(spyAddNewCity).toHaveBeenCalledWith({ newCity: cherkasyCity });
  });
});
