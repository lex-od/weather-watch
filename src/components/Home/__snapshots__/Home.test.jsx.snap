// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home component Home snapshot 1`] = `
<div>
  <div
    class="MuiBox-root css-qar1bk"
  >
    <h2
      class="MuiTypography-root MuiTypography-h2 css-905vvy-MuiTypography-root"
    >
      Weather watch
    </h2>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 css-mu8n5i-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4 css-gj1fbr-MuiGrid-root"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
        >
          <div
            class="MuiCardContent-root css-q0sxun-MuiCardContent-root"
          >
            <h5
              class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
            >
              Odessa
            </h5>
            <div
              class="iconThumb"
            >
              <img
                alt="weather icon"
                src="http://openweathermap.org/img/wn/02d@2x.png"
              />
            </div>
            <p
              class="MuiTypography-root MuiTypography-h5 css-1nool57-MuiTypography-root"
            >
              +2
              Â°, 
              Clouds
            </p>
            <p
              class="MuiTypography-root MuiTypography-subtitle1 css-ecpxie-MuiTypography-root"
            >
              Wind: 
              3
               m/s
            </p>
          </div>
          <div
            class="MuiCardActions-root MuiCardActions-spacing css-1t6e9jv-MuiCardActions-root"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiLoadingButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiLoadingButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-turab-MuiButtonBase-root-MuiButton-root-MuiLoadingButton-root"
              id=":r8:"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="RefreshIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M17.65 6.35C16.2 4.9 14.21 4 12 4c-4.42 0-7.99 3.58-7.99 8s3.57 8 7.99 8c3.73 0 6.84-2.55 7.73-6h-2.08c-.82 2.33-3.04 4-5.65 4-3.31 0-6-2.69-6-6s2.69-6 6-6c1.66 0 3.14.69 4.22 1.78L13 11h7V4l-2.35 2.35z"
                  />
                </svg>
              </span>
              Update
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-7765ua-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="DeleteIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                  />
                </svg>
              </span>
              Delete
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiBox-root css-xi606m"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeLarge MuiButton-outlinedSizeLarge MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeLarge MuiButton-outlinedSizeLarge css-9y1egq-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Add city
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;
